src/core/NgenSimulation.cpp

```diff
@@ -0,0 +1,254 @@
+
+void NgenSimulation::run_routing()
+{
+#if NGEN_WITH_ROUTING
+#if NGEN_WITH_MPI
+    if (mpi_num_procs_ > 1) {
+        int number_of_timesteps = catchment_formulation_manager_->Simulation_Time_Object->get_total_output_times();
+        std::vector<std::string> local_nexus_ids;
+        for (const auto& nexus : nexus_indexes) {
+            local_nexus_ids.push_back(nexus.first);
+        }
+        // MPI_Gather all nexus IDs into a single vector
+        std::vector<std::string> all_nexus_ids = parallel::gather_strings(local_nexus_ids, mpi_rank_, mpi_num_procs_);
+        if (mpi_rank_ == 0) {
+            // filter to only the unique IDs
+            std::sort(all_nexus_ids.begin(), all_nexus_ids.end());
+            all_nexus_ids.erase(
+                std::unique(all_nexus_ids.begin(), all_nexus_ids.end()),
+                all_nexus_ids.end()
+            );
+        }
+        // MPI_Broadcast so all processes share the nexus IDs
+        all_nexus_ids = std::move(parallel::broadcast_strings(all_nexus_ids, mpi_rank_, mpi_num_procs_));
+
+        // MPI_Reduce to collect the results from processes
+        std::vector<double> all_nexus_downflows;
+        if (mpi_rank_ == 0) {
+            all_nexus_downflows.resize(number_of_timesteps * all_nexus_ids.size(), 0.0);
+        }
+        std::unordered_map<std::string, int> all_nexus_indexes;
+        std::vector<double> local_buffer(number_of_timesteps);
+        std::vector<double> receive_buffer(number_of_timesteps, 0.0);
+        for (int i = 0; i < all_nexus_ids.size(); ++i) {
+            std::string nexus_id = all_nexus_ids[i];
+            if (nexus_indexes_.find(nexus_id) != nexus_indexes_.end() && !features.is_remote_sender_nexus(nexus_id)) {
+                // if this process has the id and receives/records data, copy the values to the buffer
+                int nexus_index = nexus_indexes_[nexus_id];
+                for (int step = 0; step < number_of_timesteps; ++step) {
+                    int offset = step * nexus_indexes_.size() + nexus_index;
+                    local_buffer[step] = nexus_downstream_flows_[offset];
+                }
+            } else {
+                // if this process does not have the id, fill with 0 to make sure it doesn't affect reduce sum
+                std::fill(local_buffer.begin(), local_buffer.end(), 0.0);
+            }
+            MPI_Reduce(local_buffer.data(), receive_buffer.data(), number_of_timesteps, MPI_DOUBLE, MPI_SUM, 0, MPI_COMM_WORLD);
+            if (mpi_rank_ == 0) {
+                // copy reduce values to a combined downflows vector
+                all_nexus_indexes[nexus_id] = i;
+                for (int step = 0; step < number_of_timesteps; ++step) {
+                    int offset = step * all_nexus_ids.size() + i;
+                    all_nexus_downflows[offset] = receive_buffer[step];
+                    receive_buffer[step] = 0.0;
+                }
+            }
+        }
+
+        if (mpi_rank_ == 0) {
```

Comment:
The NGen main function could move the collected indexes and flows since it wouldn't need to use those vectors after running t-route, but I think it'd be unwise to change NgenSimulation's members as part of this function. Maybe a local variable that is set to the member by default but will be changed to the merged values when running with MPI.